### PR TITLE
[Merged by Bors] - feat: port FieldTheory.Ratfunc

### PR DESCRIPTION
**Pull request by adomani**
I am going to try to work on this for a bit, but it looks hard.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by adomani** [comment]
I am stopping for now: feel free to step in!

**Comment by Vierkantor** [comment]
Let me look at this for an hour or two...

**Comment by adomani** [comment]
@Vierkantor this is great, thanks!  I am inching forward with removing errors!

**Comment by adomani** [comment]
Ok, I am making progress.

I unblocked a large chunk by removing the `CommRing K, IsDomain K` assumptions, when `Field K` was also present: I think that, while the `omit`table assumptions, might get omitted after parsing, but they mess up the proof, since they get entangled with the terms.

**Comment by Vierkantor** [comment]
> several `norm_cast` lemmas have a `coe` on the RHS (and two have broken proofs).

It feels like `coe_C` and `coe_X` are false positives: the `coe` on the RHS applies the bundled homs, so morally I'd say it does not coerce from one type to another.

**Comment by adomani** [comment]
Should `Ratfunc` really be `RatFunc`?  How do you change the name of the file, so that it is picked up by the automation?

**Comment by Vierkantor** [comment]
> Is there an "autofix docstrings" script?

Yes! `./scripts/fix-comments.py`. But I think we should first go through the naming conventions before we do this.

> I also have a something to say about the choice of `x` as a symbol for a generic term in `Ratfunc K` and also defining `X : Ratfunc K`... smile

Yeah, some (variable) names are not optimal but let's postpone that change until the port is complete.

> Should `Ratfunc` really be `RatFunc`? How do you change the name of the file, so that it is picked up by the automation?

I think we can just `git mv Mathlib/FieldTheory/Ratfunc.lean Mathlib/FieldTheory/RatFunc.lean`, because the file docstring includes the mathlib3 source files.

**Comment by adomani** [comment]
@Vierkantor, I "reverted" the section change: after the fact, Lean can figure out everything it seems.  It was just confusing me in the course of the proof.

As for the naming convention, I am not too familiar with the capitalization rules: can someone else (maybe you? :angel: ) help with that?

File: Mathlib/FieldTheory/Ratfunc.lean [hunk 1]
```diff
@@ -102,18 +102,27 @@ open nonZeroDivisors Polynomial
...
 the maps between `ratfunc K` and another field of fractions of `K[X]`,
 especially `fraction_ring K[X]`, are given by `is_localization.algebra_equiv`.
 -/
-structure Ratfunc : Type u where of_fraction_ring ::
-  toFractionRing : FractionRing K[X]
+structure Ratfunc : Type u where
+  /-- (implementation detail): a rational function is given by an element of the fraction field.
+
+  You should prefer `isLocalization.alg_equiv` over `of_fraction_ring`.
+  -/
```
Comment (jcommelin) [inline Mathlib/FieldTheory/Ratfunc.lean:119]: I would be strongly inclined to not add documentation during the port.
```suggestion
```

File: Mathlib/FieldTheory/Ratfunc.lean [hunk 2]
```diff
@@ -102,18 +102,27 @@ open nonZeroDivisors Polynomial
...
+  of_fraction_ring ::
+    /-- (implementation detail): a rational function gives an element of the fraction field.
+
+    You should prefer `isLocalization.alg_equiv` over `toFractionRing`.
+    -/
```
Comment (jcommelin) [inline Mathlib/FieldTheory/Ratfunc.lean:124]:
```suggestion
```

File: Mathlib/FieldTheory/Ratfunc.lean [hunk 3]
```diff
@@ -146,25 +156,30 @@ we construct a value of `P` for such elements of `ratfunc K` by setting
...
-      (let ⟨⟨c, hc⟩, mul_eq⟩ := Localization.r_iff_exists.mp h
-      mul_cancel_left_coe_nonZeroDivisors.mp mul_eq)
+protected irreducible_def liftOn {P : Sort v} (x : Ratfunc K) (f : ∀ _p _q : K[X], P)
+    (H : ∀ {p q p' q'} (_hq : q ∈ K[X]⁰) (_hq' : q' ∈ K[X]⁰), q' * p = q * p' → f p q = f p' q') :
+    P := by
+  refine Localization.liftOn (toFractionRing x) (fun p q => f p q) ?_
+  intros p p' q q' h
+  exact H q.2 q'.2 (let ⟨⟨c, hc⟩, mul_eq⟩ := Localization.r_iff_exists.mp h
+    mul_cancel_left_coe_nonZeroDivisors.mp mul_eq)
+-- porting note: the definition above was as follow
```
Comment (jcommelin) [inline Mathlib/FieldTheory/Ratfunc.lean:166]:
```suggestion
-- porting note: the definition above was as follows
```

**Comment by Vierkantor** [comment]
> As for the naming convention, I am not too familiar with the capitalization rules: can someone else (maybe you? angel ) help with that?

Doing that right now!

**Comment by Vierkantor** [comment]
Done, I think this is ready for review again.

**Review by jcommelin — APPROVED** [review]
Thanks :tada:

bors merge

**Comment by Vierkantor** [comment]
Let's try again (with @jcommelin's verbal approval)!

bors r+

**Pull request closed without merging**